File: java/list-ops/src/main/java/ListOps.java
import java.util.LinkedList;
import java.util.List;
import java.util.function.*;

public class ListOps {
    public static int length(List<Integer> list) {
        return list.size();
    }

    public static <T> List<T> map(List<T> list, UnaryOperator<T> operator) {
        List<T> mapped = new LinkedList<>();

        for (T value: list) {
            mapped.add(operator.apply(value));
        }

        return mapped;
    }

    public static <T> List<T> filter(List<T> list, Predicate<T> predicate) {
        List<T> filtered = new LinkedList<>();

        for (T value: list) {
            if (predicate.test(value)) {
                filtered.add(value);
            }
        }

        return filtered;
    }

    public static <T> List<T> reverse(List<T> list) {
        List<T> reversed = new LinkedList<>();

        for (int i = list.size() - 1; i >= 0; i--) {
            reversed.add(list.get(i));
        }

        return reversed;
    }

    public static <T> List<T> concat(List<T>... lists) {
        List<T> concatenated = new LinkedList<>();

        for (List<T> list: lists) {
            for (T value: list) {
                concatenated.add(value);
            }
        }

        return concatenated;
    }

    public static <U, T> U reduce(List<T> list, U identity, BiFunction<U, T, U> accumulator, BinaryOperator<U> combiner) {
        if (list.isEmpty()) {
            return identity;
        }

        U result = identity;
        for (T element : list) {
            result = accumulator.apply(result, element);
        }

        return result;
    }
}

File: swift/list-ops/Sources/ListOps/ListOps.swift

func foldLeft<TIn, TOut>(_ l: [TIn], accumulated acc: TOut, combine c: (TOut, TIn) -> TOut) -> TOut {
    var state = acc

    for item in l {
        state = c(state, item)
    }

    return state
} 

func foldRight<TIn, TOut>(_ l: [TIn], accumulated acc: TOut, combine c: (TIn, TOut) -> TOut) -> TOut {
    guard !l.isEmpty else { return acc }

    var state = acc

    for item in l.reversed() {
        state = c(item, state)
    }

    return state
} 

func append<T>(_ l: [T], _ r: [T]) -> [T] {
    var appended = l

    for item in r {
        appended.append(item)
    }

    return appended
} 

func concat<T>(_ lists: [T]...) -> [T] {
    var concatenated: [T] = []

    for list in lists {
        for item in list {
            concatenated.append(item)
        }
    }

    return concatenated
} 

func length<T>(_ l: [T]) -> Int {
    var length = 0;

    for _ in l {
        length += 1
    }

    return length
} 

func filter<T>(_ l: [T], f: (T) -> Bool) -> [T] {
    var mapped: [T] = [];

    for item in l {
        if f(item) {
            mapped.append(item)
        }
    }

    return mapped
} 

func map<TIn, TOut>(_ l: [TIn], f: (TIn) -> TOut) -> [TOut] {
    var mapped: [TOut] = [];

    for item in l {
        mapped.append(f(item))
    }

    return mapped
} 

func reverse<T>(_ l: [T]) -> [T] {
    var reversed: [T] = [];

    for item in l.reversed() {
        reversed.append(item)
    }

    return reversed
} 
